dp_consistency/src/report/report.py

```diff
@@ -4,112 +4,176 @@
+        self.__reach_time_series_by_edp_combination = (
+            reach_time_series_by_edp_combination
+        )
 
-    def sample_with_noise(self) -> 'MetricReport':
+    def sample_with_noise(self) -> "MetricReport":
         """
         :return: a new MetricReport where measurements have been resampled
         according to their mean and variance.
         """
         return MetricReport(
-            total_campaign_reach_time_series=list(
-                MetricReport.__sample_with_noise(measurement) for measurement in
-                self.__total_campaign_reach_time_series),
-            reach_time_series_by_edp=list(
-                list(MetricReport.__sample_with_noise(measurement)
-                     for measurement in series)
-                for series in self.__reach_time_series_by_edp))
-
-    def get_edp_measurement(self, edp: int, period: int):
-        return self.__reach_time_series_by_edp[edp][period]
-
-    def get_total_reach_measurement(self, period: int):
-        return self.__total_campaign_reach_time_series[period]
-
-    def get_num_edps(self):
-        return len(self.__reach_time_series_by_edp)
+            reach_time_series_by_edp_combination={
+                edp_comb: [
+                    MetricReport.__sample_with_noise(measurement)
+                    for measurement in self.__reach_time_series_by_edp_combination[
+                        edp_comb
+                    ]
+                ]
+                for edp_comb in self.__reach_time_series_by_edp_combination.keys()
+            }
+        )
+
+    def get_edp_comb_measurement(self, edp_comb: str, period: int):
+        return self.__reach_time_series_by_edp_combination[edp_comb][period]
+
+    def get_edp_combs(self):
+        return list(self.__reach_time_series_by_edp_combination.keys())
+
+    def get_num_edp_combs(self):
+        return len(self.__reach_time_series_by_edp_combination.keys())
 
     def get_number_of_periods(self):
-        return len(self.__total_campaign_reach_time_series)
+        return len(next(iter(self.__reach_time_series_by_edp_combination.values())))
+
+    def get_subset_relationships(self):
+        """Returns a list of tuples where first element in the tuple is the parent
+        and second element is the subset."""
+        subset_relationships = []
+        edp_combinations = list(self.__reach_time_series_by_edp_combination)
+
+        for comb1, comb2 in combinations(edp_combinations, 2):
+            if comb1.issubset(comb2):
+                subset_relationships.append((comb2, comb1))
+            elif comb2.issubset(comb1):
+                subset_relationships.append((comb1, comb2))
+        return subset_relationships
+
+    def get_cover_relationships(self):
+        """Returns covers as defined here: # https://en.wikipedia.org/wiki/Cover_(topology).
+        For each set (s_i) in the list, enumerate combinations of all sets excluding this one.
```

Comment:
you can probably get a quick performance win by only including subsets of S in the list of subsets you look for combinations of.

You can probably make it even better by introducing some memoization and noting that if A and B cover C the so must all covers(A) X covers(B)

Reply:
you may want to add a unit test for this given this is the most complex logic here.

Reply:
+1. The current algorithm may have O(n2^{n-1}) computation time, which will not scale well when there are more EDPs.
typo: union -> union